### DISCUSSION TITLE
Would transforming be possible as well?

### DISCUSSION THREAD

**Discussion by Toflar** [cognesy/instructor-php#17] [Q&A] [answered]
Hi, I just stumbled across this library by chance and I was wondering if transformation from structured to structured while applying context would also be possible?

This would be super useful for translating or transforming existing models. 
Imagine a project model like this:

```php
class Project {
    public string $title;
    public string $description;
}
```

Then you have an existing project like so:

```php
$project = new Project('A wonderful project', 'This is a superbe description meant for amateurs in the field of wonderful projects.');
```

Now what I would like to do is something like this:

```php
$translatedProject = (new Instructor())->respond(
    messages: [['role' => 'user', 'content' => 'Translate the project description into German and make sure it is translated so that an amateur would understand it. Leave the project title as is.']],
    requestModel: $project,
    responseModel: Project::class,
);

var_dump($translatedProject->title); // A wonderful project
var_dump($translatedProject->description); // Something smart in German :-)
```

Do you think that's at all possible? And would it fit this library? It's kind of also instructing and working with structured data.
I can think of many use cases:

- Translating
- Fix spelling mistakes
- Restructuring (imagine a table with rows and columns and you want to have it sorted not lexicographically but maybe by something "people would most likely emotionally connect to" or whatever is not possible using just a simple algorithm).
- Adjusting emojis used from a text that was maybe written in the US to what a reader in Japan would expect.
- ...

**Comment by ddebowczyk** [comment, marked as answer]
Apologies for late response, I've missed this question. It is possible to achieve 'structure to structure' processing.

I'm currently working on more documentation for the "language programs", the new capability that is already present in the codebase on GitHub and will be formally announced and documented in the next release.

Language programs are the way to use Instructor capabilities to execute multiple steps of LLM inference and conventional processing with structured data, also transforming the data to different structures between the steps. The idea is inspired by DSPy for Python (although initially DSPy has been focused on processing text based content with LLMs and avoiding manual prompting).

As Instructor natively works with structured data, I felt it would be good to have a modular, controllable way to process and transform data structures, but also intertwine LLM processing steps with more controllable conventional processing with regular code. This way data can be taken from external sources, transformed with (and without) LLM support, validated and finally pushed out to target destinations.

You can check the early preview here:
https://github.com/cognesy/instructor-php/blob/main/examples/02_Advanced/LanguagePrograms/run.php

It may not be very clear what is happening in the example just looking at the code, but should give you a general feeling what's possible. 

The documentation is going to be available soon and it will clarify how to use the new capabilities and what might be the applications.

**Comment by ddebowczyk** [comment]
Another example showing multi-step LLM based processing:
https://github.com/cognesy/instructor-php/blob/main/examples/02_Advanced/LanguagePrograms2/run.php

**Reply by ddebowczyk** [reply]
Even simpler example using v0.8+ API:
https://github.com/cognesy/instructor-php/blob/main/examples/02_Advanced/DataInputs/run.php
